File: push-be/internal/domain/usecase/google.go
package usecase

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"net/url"
	"strings"
	"time"

	"github.com/google/uuid"

	"push-be/internal/domain"
)

const googleCipherPurpose = "google"

type GoogleConfig struct {
	ClientID     string
	ClientSecret string
	AuthURL      string
	TokenURL     string
	GmailBeta    bool
}

func (c GoogleConfig) Configured() bool {
	return c.ClientID != "" && c.ClientSecret != ""
}

type GoogleService struct {
	store        domain.GoogleStore
	sessions     domain.SessionStore
	calendar     domain.CalendarEventStore
	applications domain.ApplicationStore
	ops          domain.OperationStore
	uow          domain.UnitOfWork
	cipher       domain.TokenCipher
	gapi         domain.GoogleClient
	cfg          GoogleConfig
}

func NewGoogleService(store domain.GoogleStore, sessions domain.SessionStore, calendar domain.CalendarEventStore, applications domain.ApplicationStore, ops domain.OperationStore, uow domain.UnitOfWork, cipher domain.TokenCipher, gapi domain.GoogleClient, cfg GoogleConfig) *GoogleService {
	return &GoogleService{
		store: store, sessions: sessions, calendar: calendar, applications: applications,
		ops: ops, uow: uow, cipher: cipher, gapi: gapi, cfg: cfg,
	}
}

func (s *GoogleService) Status(ctx context.Context, userID uuid.UUID) (*domain.GoogleStatus, error) {
	st := &domain.GoogleStatus{
		Enabled: s.cfg.Configured(), Connected: false,
		Scopes: []string{}, CalendarStatus: "DISCONNECTED",
	}
	if !s.cfg.GmailBeta {
		st.GmailStatus = "DISABLED"
	} else {
		st.GmailStatus = "DISCONNECTED"
	}
	gi, err := s.store.GetGoogleIntegration(ctx, userID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return st, nil
		}
		return nil, domain.Internal()
	}
	st.Connected = true
	st.Scopes = gi.Scopes
	st.LastSyncedAt = gi.LastSyncedAt
	if s.cfg.GmailBeta && gi.HasScope(domain.GoogleScopeGmail) {
		st.GmailStatus = "CONNECTED"
	}
	if gi.HasScope(domain.GoogleScopeCalendar) {
		st.CalendarStatus = "CONNECTED"
	}
	return st, nil
}

func (s *GoogleService) Connect(ctx context.Context, userID uuid.UUID, codeChallenge, method, redirectURI, callbackURL string) (string, string, time.Time, error) {
	if !s.cfg.Configured() {
		return "", "", time.Time{}, domain.NotConfigured("google integration is not configured")
	}
	if s.cipher == nil {
		return "", "", time.Time{}, domain.NotConfigured("token encryption is not configured")
	}
	if method != "S256" {
		return "", "", time.Time{}, domain.ValidationField("codeChallengeMethod", "must be S256")
	}
	if codeChallenge == "" {
		return "", "", time.Time{}, domain.ValidationField("codeChallenge", "required")
	}
	if redirectURI != domain.GoogleCallbackURI {
		return "", "", time.Time{}, domain.ValidationField("redirectUri", "not allowed")
	}
	state, err := randomToken()
	if err != nil {
		return "", "", time.Time{}, domain.Internal()
	}
	now := time.Now().UTC()
	rec := &domain.OAuthState{
		State: state, Provider: "google", Purpose: domain.OAuthPurposeGoogle,
		UserID: &userID, CodeChallenge: codeChallenge, RedirectURI: callbackURL,
		ExpiresAt: now.Add(domain.OAuthStateTTL), CreatedAt: now,
	}
	if err := s.sessions.SaveOAuthState(ctx, rec); err != nil {
		return "", "", time.Time{}, domain.Internal()
	}
	q := url.Values{}
	q.Set("client_id", s.cfg.ClientID)
	q.Set("redirect_uri", callbackURL)
	q.Set("response_type", "code")
	q.Set("state", state)
	q.Set("scope", domain.GoogleScopeGmail+" "+domain.GoogleScopeCalendar)
	q.Set("access_type", "offline")
	q.Set("prompt", "consent")
	return s.cfg.AuthURL + "?" + q.Encode(), state, rec.ExpiresAt, nil
}

type googleTokenPayload struct {
	AccessToken  string `json:"accessToken"`
	RefreshToken string `json:"refreshToken"`
	Scope        string `json:"scope"`
	ExpiresIn    int    `json:"expiresIn"`
}

func (s *GoogleService) HandleCallback(ctx context.Context, code, state string) (string, error) {
	if !s.cfg.Configured() {
		return "", domain.NotConfigured("google integration is not configured")
	}
	if s.cipher == nil {
		return "", domain.NotConfigured("token encryption is not configured")
	}
	rec, err := s.sessions.GetOAuthState(ctx, state)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return "", domain.Unauthenticated("invalid state")
		}
		return "", domain.Internal()
	}
	now := time.Now().UTC()
	_ = s.sessions.DeleteOAuthState(ctx, state)
	if now.After(rec.ExpiresAt) || rec.Purpose != domain.OAuthPurposeGoogle || rec.UserID == nil {
		return "", domain.Unauthenticated("state expired")
	}
	tokens, err := s.gapi.ExchangeCode(ctx, code, rec.RedirectURI)
	if err != nil {
		return "", domain.ProviderError("provider token exchange failed")
	}
	payload, _ := json.Marshal(googleTokenPayload{
		AccessToken: tokens.AccessToken, RefreshToken: tokens.RefreshToken,
		Scope: tokens.Scope, ExpiresIn: tokens.ExpiresIn,
	})
	ct, nonce, err := s.cipher.Encrypt(string(payload), *rec.UserID, googleCipherPurpose)
	if err != nil {
		return "", domain.Internal()
	}
	bundle, _ := json.Marshal(map[string]string{
		"ciphertext": base64.StdEncoding.EncodeToString(ct),
		"nonce":      base64.StdEncoding.EncodeToString(nonce),
	})
	integrationCode, err := randomToken()
	if err != nil {
		return "", domain.Internal()
	}
	if err := s.store.SaveIntegrationCode(ctx, &domain.IntegrationCode{
		Code: integrationCode, UserID: *rec.UserID, CodeChallenge: rec.CodeChallenge,
		Payload: bundle, ExpiresAt: now.Add(domain.IntegrationCodeTTL), CreatedAt: now,
	}); err != nil {
		return "", domain.Internal()
	}
	return domain.GoogleCallbackURI + "?code=" + url.QueryEscape(integrationCode), nil
}

func (s *GoogleService) Complete(ctx context.Context, userID uuid.UUID, code, codeVerifier string) (*domain.GoogleStatus, error) {
	if !s.cfg.Configured() {
		return nil, domain.NotConfigured("google integration is not configured")
	}
	if s.cipher == nil {
		return nil, domain.NotConfigured("token encryption is not configured")
	}
	rec, err := s.store.GetIntegrationCode(ctx, code)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.Unauthenticated("invalid code")
		}
		return nil, domain.Internal()
	}
	now := time.Now().UTC()
	if rec.UsedAt != nil || now.After(rec.ExpiresAt) {
		return nil, domain.Unauthenticated("code expired or already used")
	}
	if rec.UserID != userID {
		return nil, domain.Unauthenticated("code belongs to a different account")
	}
	if codeVerifier == "" || pkceS256(codeVerifier) != rec.CodeChallenge {
		return nil, domain.Unauthenticated("pkce verification failed")
	}
	var bundle struct {
		Ciphertext string `json:"ciphertext"`
		Nonce      string `json:"nonce"`
	}
	if err := json.Unmarshal(rec.Payload, &bundle); err != nil {
		return nil, domain.Internal()
	}
	ct, _ := base64.StdEncoding.DecodeString(bundle.Ciphertext)
	nonce, _ := base64.StdEncoding.DecodeString(bundle.Nonce)
	raw, err := s.cipher.Decrypt(ct, nonce, userID, googleCipherPurpose)
	if err != nil {
		return nil, domain.Internal()
	}
	var payload googleTokenPayload
	if err := json.Unmarshal([]byte(raw), &payload); err != nil {
		return nil, domain.Internal()
	}
	act, an, err := s.cipher.Encrypt(payload.AccessToken, userID, googleCipherPurpose)
	if err != nil {
		return nil, domain.Internal()
	}
	gi := &domain.GoogleIntegration{
		UserID: userID, AccessCiphertext: act, AccessNonce: an,
		Scopes: strings.Fields(payload.Scope),
		CreatedAt: now, UpdatedAt: now,
	}
	if payload.ExpiresIn > 0 {
		exp := now.Add(time.Duration(payload.ExpiresIn) * time.Second)
		gi.TokenExpiresAt = &exp
	}
	if payload.RefreshToken != "" {
		rct, rn, err := s.cipher.Encrypt(payload.RefreshToken, userID, googleCipherPurpose)
		if err != nil {
			return nil, domain.Internal()
		}
		gi.RefreshCiphertext = rct
		gi.RefreshNonce = rn
	}
	err = s.uow.Do(ctx, func(ctx context.Context) error {
		if err := s.store.MarkIntegrationCodeUsed(ctx, code, now); err != nil {
			return domain.Unauthenticated("code expired or already used")
		}
		return s.store.PutGoogleIntegration(ctx, gi)
	})
	if err != nil {
		if de, ok := err.(*domain.Error); ok {
			return nil, de
		}
		return nil, domain.Internal()
	}
	return s.Status(ctx, userID)
}

func (s *GoogleService) accessToken(ctx context.Context, gi *domain.GoogleIntegration) (string, error) {
	at, err := s.cipher.Decrypt(gi.AccessCiphertext, gi.AccessNonce, gi.UserID, googleCipherPurpose)
	if err != nil {
		return "", domain.Internal()
	}
	expired := gi.TokenExpiresAt != nil && time.Now().UTC().After(gi.TokenExpiresAt.Add(-time.Minute))
	if !expired || len(gi.RefreshCiphertext) == 0 {
		return at, nil
	}
	rt, err := s.cipher.Decrypt(gi.RefreshCiphertext, gi.RefreshNonce, gi.UserID, googleCipherPurpose)
	if err != nil {
		return "", domain.Internal()
	}
	tokens, err := s.gapi.RefreshAccessToken(ctx, rt)
	if err != nil {
		return "", domain.ProviderError("google token refresh failed")
	}
	ct, nonce, err := s.cipher.Encrypt(tokens.AccessToken, gi.UserID, googleCipherPurpose)
	if err != nil {
		return "", domain.Internal()
	}
	gi.AccessCiphertext = ct
	gi.AccessNonce = nonce
	if tokens.ExpiresIn > 0 {
		exp := time.Now().UTC().Add(time.Duration(tokens.ExpiresIn) * time.Second)
		gi.TokenExpiresAt = &exp
	}
	if err := s.store.PutGoogleIntegration(ctx, gi); err != nil {
		return "", domain.Internal()
	}
	return tokens.AccessToken, nil
}

func (s *GoogleService) Disconnect(ctx context.Context, userID uuid.UUID) error {
	gi, err := s.store.GetGoogleIntegration(ctx, userID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil
		}
		return domain.Internal()
	}
	if at, err := s.cipher.Decrypt(gi.AccessCiphertext, gi.AccessNonce, userID, googleCipherPurpose); err == nil {
		_ = s.gapi.RevokeToken(ctx, at)
	} else if len(gi.RefreshCiphertext) > 0 {
		if rt, err := s.cipher.Decrypt(gi.RefreshCiphertext, gi.RefreshNonce, userID, googleCipherPurpose); err == nil {
			_ = s.gapi.RevokeToken(ctx, rt)
		}
	}
	return s.uow.Do(ctx, func(ctx context.Context) error {
		if err := s.store.DeleteGoogleData(ctx, userID); err != nil {
			return err
		}
		return s.store.DeleteGoogleIntegration(ctx, userID)
	})
}

func (s *GoogleService) Sync(ctx context.Context, userID uuid.UUID) (*domain.Operation, error) {
	gi, err := s.store.GetGoogleIntegration(ctx, userID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.IntegrationRequired("google account is not connected")
		}
		return nil, domain.Internal()
	}
	now := time.Now().UTC()
	op := &domain.Operation{
		ID: uuid.New(), UserID: userID, Type: domain.OpGoogleSync,
		Status: domain.OpRunning, CreatedAt: now, UpdatedAt: now,
	}
	if err := s.ops.Create(ctx, op); err != nil {
		return nil, domain.Internal()
	}
	fail := func(code, msg string) (*domain.Operation, error) {
		op.Status = domain.OpFailed
		op.Error = &domain.OperationError{Code: code, Message: msg, Retryable: true}
		op.UpdatedAt = time.Now().UTC()
		_ = s.ops.Update(ctx, op)
		return op, domain.ProviderError(msg)
	}
	token, derr := s.accessToken(ctx, gi)
	if derr != nil {
		return fail(domain.CodeProviderError, "google access token unavailable")
	}
	messages := []domain.GoogleMessageMeta{}
	events := []domain.GoogleEvent{}
	historyID := gi.GmailHistoryID
	syncToken := gi.CalendarSyncToken
	if s.cfg.GmailBeta && gi.HasScope(domain.GoogleScopeGmail) {
		pageToken := ""
		for i := 0; i < 5; i++ {
			ids, next, hid, err := s.gapi.ListMessageIDs(ctx, token, pageToken)
			if err != nil {
				return fail(domain.CodeProviderError, "gmail message list failed")
			}
			for _, id := range ids {
				m, err := s.gapi.GetMessage(ctx, token, id)
				if err != nil {
					return fail(domain.CodeProviderError, "gmail message fetch failed")
				}
				messages = append(messages, *m)
			}
			if hid != "" && hid != "0" {
				historyID = hid
			}
			if next == "" {
				break
			}
			pageToken = next
		}
	}
	if gi.HasScope(domain.GoogleScopeCalendar) {
		evs, nst, err := s.gapi.ListEvents(ctx, token, syncToken)
		if errors.Is(err, domain.ErrSyncTokenInvalid) {
			evs, nst, err = s.gapi.ListEvents(ctx, token, "")
		}
		if err != nil {
			return fail(domain.CodeProviderError, "calendar event list failed")
		}
		events = evs
		if nst != "" {
			syncToken = nst
		}
	}
	gi.GmailHistoryID = historyID
	gi.CalendarSyncToken = syncToken
	gi.LastSyncedAt = &now
	gi.UpdatedAt = now
	err = s.uow.Do(ctx, func(ctx context.Context) error {
		for _, m := range messages {
			if err := s.store.UpsertGoogleMessage(ctx, &domain.GoogleMessage{
				ID: uuid.New(), UserID: userID, ExternalID: m.ID, ThreadID: m.ThreadID,
				Sender: m.From, Subject: m.Subject, Snippet: m.Snippet,
				ReceivedAt: m.ReceivedAt, CreatedAt: now,
			}); err != nil {
				return err
			}
		}
		for _, ev := range events {
			if ev.Status == "cancelled" {
				if err := s.calendar.DeleteExternalCalendarEvent(ctx, userID, ev.ID); err != nil {
					return err
				}
				continue
			}
			tz := ev.TimeZone
			if tz == "" {
				tz = "UTC"
			}
			endsAt := ev.EndsAt
			if !endsAt.After(ev.StartsAt) {
				endsAt = ev.StartsAt.Add(time.Hour)
			}
			extID := ev.ID
			if err := s.calendar.UpsertExternalCalendarEvent(ctx, &domain.CalendarEvent{
				ID: uuid.New(), UserID: userID, Revision: 1, Type: domain.CalCustom,
				Title: ev.Summary, StartsAt: ev.StartsAt, EndsAt: endsAt, TimeZone: tz,
				Source: domain.EventSourceGoogle, ExternalID: &extID,
				CreatedAt: now, UpdatedAt: now,
			}); err != nil {
				return err
			}
		}
		if err := s.store.PutGoogleIntegration(ctx, gi); err != nil {
			return err
		}
		res, _ := json.Marshal(map[string]any{
			"messages": len(messages), "events": len(events), "lastSyncedAt": now,
		})
		op.Status = domain.OpSucceeded
		op.Result = &domain.OperationResult{Kind: string(domain.OpGoogleSync), Value: res}
		op.UpdatedAt = now
		return s.ops.Update(ctx, op)
	})
	if err != nil {
		return fail(domain.CodeInternal, "google sync persistence failed")
	}
	return op, nil
}

func (s *GoogleService) requireConnection(ctx context.Context, userID uuid.UUID) error {
	_, err := s.store.GetGoogleIntegration(ctx, userID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return domain.IntegrationRequired("google account is not connected")
		}
		return domain.Internal()
	}
	return nil
}

func (s *GoogleService) ListMessages(ctx context.Context, userID uuid.UUID, applicationID *uuid.UUID, page domain.PageRequest) (domain.Page[domain.GoogleMessage], error) {
	if err := s.requireConnection(ctx, userID); err != nil {
		return domain.Page[domain.GoogleMessage]{}, err
	}
	limit := page.EffectiveLimit()
	items, err := s.store.ListGoogleMessages(ctx, userID, applicationID, page, limit+1)
	if err != nil {
		return domain.Page[domain.GoogleMessage]{}, domain.Internal()
	}
	return domain.NewPage(items, limit, func(m domain.GoogleMessage) domain.Cursor {
		return domain.Cursor{CreatedAt: m.CreatedAt, ID: m.ID}
	}), nil
}

func (s *GoogleService) ListEvents(ctx context.Context, userID uuid.UUID, from, to *time.Time, page domain.PageRequest) (domain.Page[domain.CalendarEvent], error) {
	if err := s.requireConnection(ctx, userID); err != nil {
		return domain.Page[domain.CalendarEvent]{}, err
	}
	src := domain.EventSourceGoogle
	limit := page.EffectiveLimit()
	items, err := s.calendar.ListCalendarEvents(ctx, userID, domain.CalendarEventFilter{
		From: from, To: to, Source: &src, Page: page,
	}, limit+1)
	if err != nil {
		return domain.Page[domain.CalendarEvent]{}, domain.Internal()
	}
	return domain.NewPage(items, limit, func(e domain.CalendarEvent) domain.Cursor {
		return domain.Cursor{CreatedAt: e.CreatedAt, ID: e.ID}
	}), nil
}

func (s *GoogleService) LinkMessage(ctx context.Context, userID, messageID, applicationID uuid.UUID) (*domain.GoogleMessage, error) {
	if err := s.requireConnection(ctx, userID); err != nil {
		return nil, err
	}
	m, err := s.store.GetGoogleMessage(ctx, userID, messageID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NotFound()
		}
		return nil, domain.Internal()
	}
	if _, err := s.applications.Get(ctx, userID, applicationID); err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NotFound()
		}
		return nil, domain.Internal()
	}
	if err := s.store.LinkGoogleMessage(ctx, userID, messageID, applicationID); err != nil {
		return nil, domain.Internal()
	}
	m.ApplicationID = &applicationID
	return m, nil
}

File: push-be/app/domain/services/google.py
from __future__ import annotations
import base64
import json
from datetime import datetime, timedelta, timezone
from typing import Optional
from urllib.parse import urlencode
from uuid import UUID, uuid4

from app.db import DB, NotFoundError
from app.domain import entities as ent
from app.domain.errors import (
    DomainError, integration_required, internal, not_configured, not_found,
    provider_error, unauthenticated, validation_field,
)
from app.infrastructure.crypto import random_token, verify_pkce
from app.infrastructure.google_client import SyncTokenInvalid
from app.infrastructure.store_auth import SessionStore
from app.infrastructure.store_calendar import CalendarEventStore
from app.infrastructure.store_integrations import IntegrationStore
from app.infrastructure.store_applications import ApplicationStore
from app.infrastructure.store_operations import OperationStore
from app.jsonutil import to_jsonable

GOOGLE_CIPHER_PURPOSE = "google"
GOOGLE_SCOPE_GMAIL = "https://www.googleapis.com/auth/gmail.readonly"
GOOGLE_SCOPE_CALENDAR = "https://www.googleapis.com/auth/calendar.events"
GOOGLE_CALLBACK_URI = "push://integrations/google/callback"
OAUTH_STATE_TTL = timedelta(minutes=10)
INTEGRATION_CODE_TTL = timedelta(seconds=60)


def _now() -> datetime:
    return datetime.now(timezone.utc)


class GoogleService:
    def __init__(self, db: DB, cipher, gapi, client_id: str,
                 client_secret: str, auth_url: str, gmail_beta: bool):
        self.db = db
        self.store = IntegrationStore(db)
        self.sessions = SessionStore(db)
        self.calendar = CalendarEventStore(db)
        self.applications = ApplicationStore(db)
        self.ops = OperationStore(db)
        self.cipher = cipher
        self.gapi = gapi
        self.client_id = client_id
        self.client_secret = client_secret
        self.auth_url = auth_url
        self.gmail_beta = gmail_beta

    @property
    def configured(self) -> bool:
        return bool(self.client_id and self.client_secret)

    def _check(self) -> None:
        if not self.configured:
            raise not_configured("google integration is not configured")
        if self.cipher is None:
            raise not_configured("token encryption is not configured")

    async def status(self, user_id: UUID) -> ent.GoogleStatus:
        st = ent.GoogleStatus(
            enabled=self.configured, connected=False,
            gmail_status="DISCONNECTED" if self.gmail_beta else "DISABLED",
            calendar_status="DISCONNECTED")
        try:
            gi = await self.store.get_google_integration(user_id)
        except NotFoundError:
            return st
        except Exception:
            raise internal()
        st.connected = True
        st.scopes = gi.scopes
        st.last_synced_at = gi.last_synced_at
        if self.gmail_beta and gi.has_scope(GOOGLE_SCOPE_GMAIL):
            st.gmail_status = "CONNECTED"
        if gi.has_scope(GOOGLE_SCOPE_CALENDAR):
            st.calendar_status = "CONNECTED"
        return st

    async def connect(self, user_id: UUID, code_challenge: str, method: str,
                      redirect_uri: str, callback_url: str):
        self._check()
        if method != "S256":
            raise validation_field("codeChallengeMethod", "must be S256")
        if not code_challenge:
            raise validation_field("codeChallenge", "required")
        if redirect_uri != GOOGLE_CALLBACK_URI:
            raise validation_field("redirectUri", "not allowed")
        state = random_token()
        now = _now()
        try:
            await self.sessions.save_oauth_state(ent.OAuthState(
                state=state, provider="google", purpose="GOOGLE",
                user_id=user_id, code_challenge=code_challenge,
                redirect_uri=callback_url,
                expires_at=now + OAUTH_STATE_TTL, created_at=now))
        except Exception:
            raise internal()
        q = urlencode({
            "client_id": self.client_id,
            "redirect_uri": callback_url,
            "response_type": "code",
            "state": state,
            "scope": GOOGLE_SCOPE_GMAIL + " " + GOOGLE_SCOPE_CALENDAR,
            "access_type": "offline",
            "prompt": "consent"})
        return self.auth_url + "?" + q, state, now + OAUTH_STATE_TTL

    async def handle_callback(self, code: str, state: str) -> str:
        self._check()
        try:
            rec = await self.sessions.get_oauth_state(state)
        except NotFoundError:
            raise unauthenticated("invalid state")
        except Exception:
            raise internal()
        now = _now()
        await self.sessions.delete_oauth_state(state)
        if (now > rec.expires_at or rec.purpose != "GOOGLE"
                or rec.user_id is None):
            raise unauthenticated("state expired")
        try:
            tokens = await self.gapi.exchange_code(code, rec.redirect_uri)
        except Exception:
            raise provider_error("provider token exchange failed")
        payload = json.dumps({
            "accessToken": tokens.access_token,
            "refreshToken": tokens.refresh_token,
            "scope": tokens.scope,
            "expiresIn": tokens.expires_in})
        try:
            ct, nonce = self.cipher.encrypt(
                payload, rec.user_id, GOOGLE_CIPHER_PURPOSE)
        except Exception:
            raise internal()
        bundle = {
            "ciphertext": base64.b64encode(ct).decode(),
            "nonce": base64.b64encode(nonce).decode()}
        integration_code = random_token()
        try:
            await self.store.save_integration_code(ent.IntegrationCode(
                code=integration_code, user_id=rec.user_id,
                code_challenge=rec.code_challenge, payload=bundle,
                expires_at=now + INTEGRATION_CODE_TTL, created_at=now))
        except Exception:
            raise internal()
        return GOOGLE_CALLBACK_URI + "?code=" + integration_code

    async def complete(self, user_id: UUID, code: str,
                       code_verifier: str) -> ent.GoogleStatus:
        self._check()
        try:
            rec = await self.store.get_integration_code(code)
        except NotFoundError:
            raise unauthenticated("invalid code")
        except Exception:
            raise internal()
        now = _now()
        if rec.used_at is not None or now > rec.expires_at:
            raise unauthenticated("code expired or already used")
        if rec.user_id != user_id:
            raise unauthenticated("code belongs to a different account")
        if not code_verifier or not verify_pkce(code_verifier, rec.code_challenge):
            raise unauthenticated("pkce verification failed")
        try:
            ct = base64.b64decode(rec.payload["ciphertext"])
            nonce = base64.b64decode(rec.payload["nonce"])
            raw = self.cipher.decrypt(
                ct, nonce, user_id, GOOGLE_CIPHER_PURPOSE)
            payload = json.loads(raw)
        except Exception:
            raise internal()
        try:
            act, an = self.cipher.encrypt(
                payload["accessToken"], user_id, GOOGLE_CIPHER_PURPOSE)
        except Exception:
            raise internal()
        gi = ent.GoogleIntegration(
            user_id=user_id, access_ciphertext=act, access_nonce=an,
            scopes=payload.get("scope", "").split(),
            created_at=now, updated_at=now)
        if payload.get("expiresIn"):
            gi.token_expires_at = now + timedelta(
                seconds=payload["expiresIn"])
        if payload.get("refreshToken"):
            rct, rn = self.cipher.encrypt(
                payload["refreshToken"], user_id, GOOGLE_CIPHER_PURPOSE)
            gi.refresh_ciphertext = rct
            gi.refresh_nonce = rn

        async def work():
            try:
                await self.store.mark_integration_code_used(code, now)
            except Exception:
                raise unauthenticated("code expired or already used")
            await self.store.put_google_integration(gi)

        try:
            await self.db.do(work)
        except DomainError:
            raise
        except Exception:
            raise internal()
        return await self.status(user_id)

    async def _access_token(self, gi: ent.GoogleIntegration) -> str:
        try:
            at = self.cipher.decrypt(
                gi.access_ciphertext, gi.access_nonce, gi.user_id,
                GOOGLE_CIPHER_PURPOSE)
        except Exception:
            raise internal()
        expired = (gi.token_expires_at is not None
                   and _now() > gi.token_expires_at - timedelta(minutes=1))
        if not expired or not gi.refresh_ciphertext:
            return at
        try:
            rt = self.cipher.decrypt(
                gi.refresh_ciphertext, gi.refresh_nonce, gi.user_id,
                GOOGLE_CIPHER_PURPOSE)
            tokens = await self.gapi.refresh_access_token(rt)
        except Exception:
            raise provider_error("google token refresh failed")
        ct, nonce = self.cipher.encrypt(
            tokens.access_token, gi.user_id, GOOGLE_CIPHER_PURPOSE)
        gi.access_ciphertext = ct
        gi.access_nonce = nonce
        if tokens.expires_in > 0:
            gi.token_expires_at = _now() + timedelta(
                seconds=tokens.expires_in)
        try:
            await self.store.put_google_integration(gi)
        except Exception:
            raise internal()
        return tokens.access_token

    async def disconnect(self, user_id: UUID) -> None:
        try:
            gi = await self.store.get_google_integration(user_id)
        except NotFoundError:
            return
        except Exception:
            raise internal()
        try:
            at = self.cipher.decrypt(
                gi.access_ciphertext, gi.access_nonce, user_id,
                GOOGLE_CIPHER_PURPOSE)
            await self.gapi.revoke_token(at)
        except Exception:
            if gi.refresh_ciphertext:
                try:
                    rt = self.cipher.decrypt(
                        gi.refresh_ciphertext, gi.refresh_nonce, user_id,
                        GOOGLE_CIPHER_PURPOSE)
                    await self.gapi.revoke_token(rt)
                except Exception:
                    pass

        async def work():
            await self.store.delete_google_data(user_id)
            await self.store.delete_google_integration(user_id)

        try:
            await self.db.do(work)
        except Exception:
            raise internal()

    async def sync(self, user_id: UUID) -> ent.Operation:
        try:
            gi = await self.store.get_google_integration(user_id)
        except NotFoundError:
            raise integration_required("google account is not connected")
        except Exception:
            raise internal()
        now = _now()
        op = ent.Operation(
            id=uuid4(), user_id=user_id, type=ent.OP_GOOGLE_SYNC,
            status=ent.OP_RUNNING, created_at=now, updated_at=now)
        try:
            await self.ops.create(op)
        except Exception:
            raise internal()

        async def fail(code: str, msg: str):
            op.status = ent.OP_FAILED
            op.error = ent.OperationError(code=code, message=msg,
                                          retryable=True)
            op.updated_at = _now()
            try:
                await self.ops.update(op)
            except Exception:
                pass
            raise provider_error(msg)

        try:
            token = await self._access_token(gi)
        except DomainError:
            await fail("PROVIDER_ERROR", "google access token unavailable")

        messages = []
        events = []
        history_id = gi.gmail_history_id
        sync_token = gi.calendar_sync_token
        if self.gmail_beta and gi.has_scope(GOOGLE_SCOPE_GMAIL):
            page_token = ""
            for _ in range(5):
                try:
                    ids, nxt, hid = await self.gapi.list_message_ids(
                        token, page_token)
                except Exception:
                    await fail("PROVIDER_ERROR", "gmail message list failed")
                for mid in ids:
                    try:
                        m = await self.gapi.get_message(token, mid)
                    except Exception:
                        await fail("PROVIDER_ERROR", "gmail message fetch failed")
                    messages.append(m)
                if hid and hid != "0":
                    history_id = hid
                if not nxt:
                    break
                page_token = nxt
        if gi.has_scope(GOOGLE_SCOPE_CALENDAR):
            try:
                events, nst = await self.gapi.list_events(token, sync_token)
            except SyncTokenInvalid:
                try:
                    events, nst = await self.gapi.list_events(token, "")
                except Exception:
                    await fail("PROVIDER_ERROR", "calendar event list failed")
            except Exception:
                await fail("PROVIDER_ERROR", "calendar event list failed")
            if nst:
                sync_token = nst
        gi.gmail_history_id = history_id
        gi.calendar_sync_token = sync_token
        gi.last_synced_at = now
        gi.updated_at = now

        async def work():
            for m in messages:
                await self.store.upsert_google_message(ent.GoogleMessage(
                    id=uuid4(), user_id=user_id, external_id=m.id,
                    thread_id=m.thread_id, sender=m.from_,
                    subject=m.subject, snippet=m.snippet,
                    received_at=m.received_at, created_at=now))
            for ev in events:
                if ev.status == "cancelled":
                    await self.calendar.delete_external(user_id, ev.id)
                    continue
                tz = ev.time_zone or "UTC"
                ends = ev.ends_at
                if ends <= ev.starts_at:
                    ends = ev.starts_at + timedelta(hours=1)
                await self.calendar.upsert_external(ent.CalendarEvent(
                    id=uuid4(), user_id=user_id, revision=1,
                    type="CUSTOM", title=ev.summary,
                    starts_at=ev.starts_at, ends_at=ends, time_zone=tz,
                    source="GOOGLE", external_id=ev.id,
                    created_at=now, updated_at=now))
            await self.store.put_google_integration(gi)
            op.status = ent.OP_SUCCEEDED
            op.result = ent.OperationResult(
                kind=ent.OP_GOOGLE_SYNC,
                value={"messages": len(messages), "events": len(events),
                       "lastSyncedAt": to_jsonable(now)})
            op.updated_at = now
            await self.ops.update(op)

        try:
            await self.db.do(work)
        except Exception:
            await fail("INTERNAL", "google sync persistence failed")
        return op

    async def _require_connection(self, user_id: UUID) -> None:
        try:
            await self.store.get_google_integration(user_id)
        except NotFoundError:
            raise integration_required("google account is not connected")
        except Exception:
            raise internal()

    async def list_messages(self, user_id: UUID,
                            application_id: Optional[UUID], page):
        await self._require_connection(user_id)
        try:
            return await self.store.list_google_messages(
                user_id, application_id, page)
        except Exception:
            raise internal()

    async def list_events(self, user_id: UUID, from_: Optional[datetime],
                          to: Optional[datetime], page):
        await self._require_connection(user_id)
        try:
            return await self.calendar.list(
                user_id, None, from_, to, "GOOGLE", page)
        except Exception:
            raise internal()

    async def link_message(self, user_id: UUID, message_id: UUID,
                           application_id: UUID) -> ent.GoogleMessage:
        await self._require_connection(user_id)
        try:
            m = await self.store.get_google_message(user_id, message_id)
        except NotFoundError:
            raise not_found()
        except Exception:
            raise internal()
        try:
            await self.applications.get(user_id, application_id)
        except NotFoundError:
            raise not_found()
        except Exception:
            raise internal()
        try:
            await self.store.link_google_message(user_id, message_id, application_id)
        except Exception:
            raise internal()
        m.application_id = application_id
        return m
